File: examples/typescript/task_manager/agent.ts
/**
 * Task Manager - Multi-agent ADK application in TypeScript.
 *
 * Demonstrates:
 * - LlmAgent with FunctionTool and Zod schemas
 * - SequentialAgent pipeline (planner -> executor -> reviewer)
 * - Callbacks for guardrails
 * - State management via outputKey and instruction templating
 */

import {
  LlmAgent,
  SequentialAgent,
  FunctionTool,
  type CallbackContext,
} from '@google/adk';
import type { Content } from '@google/genai';
import { z } from 'zod';

// --- Tools ---

const createTask = new FunctionTool({
  name: 'create_task',
  description: 'Create a new task with a title, description, and priority.',
  parameters: z.object({
    title: z.string().describe('Short task title'),
    description: z.string().describe('Detailed task description'),
    priority: z.enum(['low', 'medium', 'high', 'critical']).describe('Task priority level'),
  }),
  execute: async ({ title, description, priority }) => {
    const id = `TASK-${Math.floor(Math.random() * 9000) + 1000}`;
    return {
      id,
      title,
      description,
      priority,
      status: 'created',
      created_at: new Date().toISOString(),
    };
  },
});

const listTasks = new FunctionTool({
  name: 'list_tasks',
  description: 'List all tasks, optionally filtered by status.',
  parameters: z.object({
    status: z.enum(['all', 'created', 'in_progress', 'done'])
      .optional()
      .describe('Filter tasks by status'),
  }),
  execute: async ({ status }) => {
    const tasks = [
      { id: 'TASK-001', title: 'Set up CI/CD', priority: 'high', status: 'in_progress' },
      { id: 'TASK-002', title: 'Write unit tests', priority: 'medium', status: 'created' },
      { id: 'TASK-003', title: 'Update docs', priority: 'low', status: 'done' },
    ];
    const filtered = status && status !== 'all'
      ? tasks.filter(t => t.status === status)
      : tasks;
    return { tasks: filtered, total: filtered.length };
  },
});

const estimateEffort = new FunctionTool({
  name: 'estimate_effort',
  description: 'Estimate the effort required for a task in hours.',
  parameters: z.object({
    task_description: z.string().describe('Description of the task to estimate'),
    complexity: z.enum(['simple', 'moderate', 'complex']).describe('Task complexity'),
  }),
  execute: async ({ task_description, complexity }) => {
    const baseHours: Record<string, number> = { simple: 2, moderate: 8, complex: 24 };
    const hours = baseHours[complexity] || 8;
    return {
      task: task_description,
      complexity,
      estimated_hours: hours,
      confidence: complexity === 'simple' ? 'high' : complexity === 'moderate' ? 'medium' : 'low',
    };
  },
});

// --- Callbacks ---

function beforeAgentCallback(callbackContext: CallbackContext): Content | undefined {
  console.log(`[Callback] Entering agent: ${callbackContext.agentName}`);
  return undefined; // Proceed normally
}

function beforeModelCallback({
  context,
  request,
}: {
  context: CallbackContext;
  request: any;
}): any | undefined {
  console.log(`[Callback] Before model call for: ${context.agentName}`);

  // Safety guardrail: block if "BLOCK" keyword detected
  const lastMessage = request.contents?.at(-1)?.parts?.[0]?.text ?? '';
  if (lastMessage.toUpperCase().includes('BLOCK')) {
    return {
      content: {
        role: 'model' as const,
        parts: [{ text: 'Request blocked by safety guardrail.' }],
      },
    };
  }

  return undefined; // Proceed normally
}

// --- Agents ---

// Stage 1: Planner - analyzes request and creates tasks
const planner = new LlmAgent({
  name: 'planner',
  model: 'gemini-2.0-flash',
  instruction: `You are a project planner. When given a project or goal:
1. Break it down into individual tasks using create_task
2. Use estimate_effort to estimate each task
3. Summarize the plan with tasks and total estimated hours

Always create at least 2 tasks for any project.`,
  description: 'Breaks down projects into tasks with effort estimates',
  tools: [createTask, estimateEffort],
  outputKey: 'project_plan',
  beforeAgentCallback,
  beforeModelCallback,
});

// Stage 2: Executor - organizes and prioritizes
const executor = new LlmAgent({
  name: 'executor',
  model: 'gemini-2.0-flash',
  instruction: `You are a project executor. Based on this project plan:

{project_plan}

1. List all existing tasks using list_tasks
2. Organize the new tasks by priority (critical > high > medium > low)
3. Create a recommended execution order
4. Identify any dependencies between tasks`,
  description: 'Organizes tasks and creates execution schedule',
  tools: [listTasks],
  outputKey: 'execution_plan',
  beforeAgentCallback,
});

// Stage 3: Reviewer - reviews the plan
const reviewer = new LlmAgent({
  name: 'reviewer',
  model: 'gemini-2.0-flash',
  instruction: `You are a project reviewer. Review this execution plan:

{execution_plan}

Evaluate:
1. Are priorities correctly assigned?
2. Is the effort estimation reasonable?
3. Are there missing tasks or dependencies?

Score 1-10 and provide brief feedback. If score >= 8, respond with "APPROVED".`,
  description: 'Reviews and scores the project plan',
  outputKey: 'review_result',
  beforeAgentCallback,
});

// Root agent: Sequential pipeline
export const rootAgent = new SequentialAgent({
  name: 'task_manager_pipeline',
  description: 'Task manager that plans, organizes, and reviews project tasks',
  subAgents: [planner, executor, reviewer],
});

File: examples/python/research_agent/agent.py
"""Research Assistant - Multi-agent ADK application.

Demonstrates:
- LlmAgent with custom tools
- SequentialAgent pipeline (researcher -> writer -> reviewer)
- Callbacks for guardrails
- State management via output_key and instruction templating
"""

from google.adk.agents import Agent, SequentialAgent
from .tools import search_articles, get_topic_stats, format_citation


def before_agent_callback(callback_context, **kwargs):
    """Log agent entry and enforce topic restrictions."""
    agent_name = callback_context.agent_name
    print(f"[Callback] Entering agent: {agent_name}")
    return None  # Proceed normally


def before_model_callback(callback_context, llm_request, **kwargs):
    """Safety guardrail: block requests containing harmful content markers."""
    # Check the last user message for blocked content
    if llm_request and hasattr(llm_request, "contents"):
        for content in (llm_request.contents or []):
            if hasattr(content, "parts"):
                for part in (content.parts or []):
                    if hasattr(part, "text") and part.text:
                        if "BLOCKED" in part.text.upper():
                            from google.genai.types import Content, Part

                            return Content(
                                role="model",
                                parts=[Part(text="Request blocked by safety guardrail.")],
                            )
    return None  # Proceed normally


# Stage 1: Research agent - gathers information
researcher = Agent(
    name="researcher",
    model="gemini-2.0-flash",
    instruction="""You are a research specialist. When given a topic:
1. Use search_articles to find relevant papers
2. Use get_topic_stats to get publication statistics
3. Summarize your findings concisely

Always include specific data from the tools in your summary.""",
    description="Gathers research articles and statistics on a topic",
    tools=[search_articles, get_topic_stats],
    output_key="research_findings",
    before_agent_callback=before_agent_callback,
    before_model_callback=before_model_callback,
)

# Stage 2: Writer agent - creates a summary report
writer = Agent(
    name="writer",
    model="gemini-2.0-flash",
    instruction="""You are a technical writer. Based on the research findings below,
write a concise research summary report (3-4 paragraphs).

Research findings:
{research_findings}

Include:
- Overview of the topic and its significance
- Key findings from the articles
- Publication trends and statistics
- Recommended areas for further study

Use format_citation for any article references.""",
    description="Writes a structured research summary from findings",
    tools=[format_citation],
    output_key="draft_report",
    before_agent_callback=before_agent_callback,
)

# Stage 3: Reviewer agent - provides quality feedback
reviewer = Agent(
    name="reviewer",
    model="gemini-2.0-flash",
    instruction="""You are a quality reviewer. Review this draft report:

{draft_report}

Evaluate on:
1. Accuracy - Are the facts and citations correct?
2. Clarity - Is it easy to understand?
3. Completeness - Does it cover key aspects?

Provide a quality score (1-10) and brief feedback.
If the score is 8 or above, approve with "APPROVED".
Otherwise provide specific improvements needed.""",
    description="Reviews and scores the draft report",
    output_key="review_result",
    before_agent_callback=before_agent_callback,
)

# Root agent: Sequential pipeline
root_agent = SequentialAgent(
    name="research_pipeline",
    description="Research assistant that finds, writes, and reviews topic summaries",
    sub_agents=[researcher, writer, reviewer],
)
